fp/predicate/predicate.go

```diff
@@ -13,8 +13,22 @@
 // limitations under the License.
 package predicate
 
+import (
+	F "github.com/terraform-provider-hpcr/fp/function"
+)
+
 func Not[A any](predicate func(A) bool) func(A) bool {
 	return func(a A) bool {
 		return !predicate((a))
 	}
 }
+
+// ContraMap creates a predicate from an existing predicate given a mapping functio
+func ContraMap[A, B any](f func(B) A) func(func(A) bool) func(B) bool {
```

Comment:
This ContraMap is in the predicate folder, so it's the version for a predicate function. Hence the return value. Versions for other Monads do not exist since we do not need them in this project.